Std/Data/List/Perm.lean

```diff
@@ -0,0 +1,168 @@
+/-
+Copyright (c) 2015 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Leonardo de Moura, Jeremy Avigad, Mario Carneiro
+-/
+
+import Std.Data.List.Basic
+import Std.Data.List.Lemmas
+
+/-!
+# List Permutations
+
+This file introduces the `List.Perm` relation, which is true if two lists are permutations of one
+another.
+
+## Notation
+
+The notation `~` is used for permutation equivalence.
+-/
+
+namespace List
+
+/-- `Perm l₁ l₂` or `l₁ ~ l₂` asserts that `l₁` and `l₂` are Permutations
+  of each other. This is defined by induction using pairwise swaps. -/
+inductive Perm {α} : List α → List α → Prop
+  /-- The empty list is a permutation of itself. -/
+  | nil   : Perm [] []
+  /-- Consing preserves the `Perm` relation. -/
+  | cons  : ∀ (x : α) {l₁ l₂ : List α}, Perm l₁ l₂ → Perm (x::l₁) (x::l₂)
+  /-- Lists with the first two elements swapped are permutations of each other. -/
+  | swap  : ∀ (x y : α) (l : List α), Perm (y::x::l) (x::y::l)
+  /-- The `Perm` relation is transitive. -/
+  | trans : ∀ {l₁ l₂ l₃ : List α}, Perm l₁ l₂ → Perm l₂ l₃ → Perm l₁ l₃
+
+open Perm
+
+/-- Notation for two lists being permutations of each other. See also `Perm`. -/
+infixl:50 " ~ " => Perm
```

Comment:
I don't think this should be a global notation.

Reply:
Consider using `Setoid` instead?

Reply:
I think it is fine to have it (and `<+`, `<+:` etc) be scoped in `List`.

Reply:
Done.